---
title: "gs"
output: html_document
fig_width: 2
fig_height: 2
---


**The deconv_ensemble app allow users to perform ensemble deconvolution as well as other deconvolution methods.**

- *Analyze* your data in the "Analysis" tab.

# Instructions


Code can be found on github: 

Please post [issues on github](https://github.com/randel/EnsDeconv), and feel free to contribute by forking and submitting development branches.

To run this app locally on your machine, download R or RStudio and run the following commands once to set up the environment:

To get detailed description of how to choose input, click:
![](../figs/help.png)



## Input Data 

You may use this app by

1. Upload your own bulk data that is either
    i. RNA-seq data 
    ii. Microarray data
    iii. DNA methylation data

### Bulk Data and Reference Format 

- Can be a .CSV *comma-separated-value* file (you may export from Excel), .rds, and .txt file
- File must have a header row.
- First/Left-hand column(s) must be gene identifiers.


#### Count or Expression Data

- Each row denotes a gene, each column denotes a sample.

<img src="../figs/bulk_example.png" width="700">

Analysis: When raw counts are uploaded, the data is then analyzed by the app. The app allow user to choose data transformation and data scaling approaches.

The gene names between bulk data and reference data should be consistent or have intersection.

#### Meta Data for Reference Data
If you choose to upload reference data instead of using in-house datasets, you will also need to upload metadata for the reference data. The metadata should include sample names for the reference (column names of the reference) and deconvolution clusters, such as cell types.

### Deconvolution important factors
For a detailed description of important factors in deconvolution, please refer to our original [paper](https://academic.oup.com/bioinformatics/article-abstract/38/11/3004/6570586?redirectedFrom=fulltext).


## Output 
You will find a boxplot, heatmap, and table displaying the estimated fraction from EnsDeconv.

Once you have chose all the paramters, click "Run" to start analysis.

You can also download estimated cell type fractions (including EnsDeconv and all other scenarios) by clicking “Download Results”:



![](../figs/download.png)



